// _@bone_iot-gateway@2.0.3@@bone/iot-gateway/src/appkey-form.js
export default function({ message, resolve, reject }){
    const options = {
            title: message,
            formConfig: [
                {
                    type: 'input',
                    name: 'appKey',
                    label: 'appKey：',
                    options: {
                        rules: [{
                            required: true,
                            message: 'appKey不能为空'
                        }]
                    }
                }, {
                    type: 'input',
                    name: 'appSecret',
                    label: 'appSecret：',
                    options: {
                        rules: [{
                            required: true,
                            message: 'appSecret不能为空'
                        }]
                    }
                }
            ],
            onOk: data => {
                fetch('/set-app-key', {
                    method: 'post',
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(data)
                }).then(response => {
                    if (response.status === 200) {
                        resolve();
                    }
                }).catch(reject);
            },
            onCancel: () => {
                reject();
            },
            onClose: () => {
                reject();
            }
        },
        { $boneStatics = {} } = window,
        dialog = $boneStatics.FormDialog;

    dialog && dialog(options);
}
