// _@bone_iot-gateway@2.0.3@@bone/iot-gateway/lib/appkey-form.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

exports.default = function (_ref) {
    var message = _ref.message,
        resolve = _ref.resolve,
        reject = _ref.reject;
    var options = {
        title: message,
        formConfig: [{
            type: 'input',
            name: 'appKey',
            label: 'appKey：',
            options: {
                rules: [{
                    required: true,
                    message: 'appKey不能为空'
                }]
            }
        }, {
            type: 'input',
            name: 'appSecret',
            label: 'appSecret：',
            options: {
                rules: [{
                    required: true,
                    message: 'appSecret不能为空'
                }]
            }
        }],
        onOk: function onOk(data) {
            fetch('/set-app-key', {
                method: 'post',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: (0, _stringify2.default)(data)
            }).then(function (response) {
                if (response.status === 200) {
                    resolve();
                }
            }).catch(reject);
        },
        onCancel: function onCancel() {
            reject();
        },
        onClose: function onClose() {
            reject();
        }
    },
        _window = window,
        _window$$boneStatics = _window.$boneStatics,
        $boneStatics = _window$$boneStatics === undefined ? {} : _window$$boneStatics,
        dialog = $boneStatics.FormDialog;


    dialog && dialog(options);
};

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
