// _@bone_web-router@0.1.15@@bone/web-router/lib/router/Router.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require("babel-runtime/core-js/object/get-prototype-of");

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require("babel-runtime/helpers/possibleConstructorReturn");

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require("babel-runtime/helpers/inherits");

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _reactRouterRedux = require("react-router-redux");

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Switch = require("./Switch");

var _Switch2 = _interopRequireDefault(_Switch);

var _Route = require("./Route");

var _Route2 = _interopRequireDefault(_Route);

var _Stacks = require("./Stacks");

var _Stacks2 = _interopRequireDefault(_Stacks);

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var layoutModelStacks = new _Stacks2.default();
var pageModelStacks = new _Stacks2.default();

// 将routes按照layout进行分组
function classifyRoutes(routes) {
    var layoutRoutes = [];

    routes.forEach(function (route, index) {
        route.index = index;
        var layout = route.layout || route.component.Layout;
        var layoutIndex = layoutRoutes.findIndex(function (layoutRoute) {
            return layoutRoute.layout === layout;
        });
        if (layoutIndex === -1) {
            layoutRoutes.push({
                layout: layout,
                routes: [route]
            });
        } else {
            layoutRoutes[layoutIndex].routes.push(route);
        }
    });

    return layoutRoutes;
}

/**
 * 
 */

var Router = function (_Component) {
    (0, _inherits3.default)(Router, _Component);

    function Router() {
        (0, _classCallCheck3.default)(this, Router);
        return (0, _possibleConstructorReturn3.default)(this, (Router.__proto__ || (0, _getPrototypeOf2.default)(Router)).apply(this, arguments));
    }

    (0, _createClass3.default)(Router, [{
        key: "getChildContext",
        value: function getChildContext() {
            return {
                layoutStacks: layoutModelStacks,
                pageStacks: pageModelStacks
            };
        }
    }, {
        key: "componentWillMount",
        value: function componentWillMount() {
            this.setState({
                layoutRoutes: classifyRoutes(this.props.routes)
            });
        }

        // 渲染页面

    }, {
        key: "renderRoutes",
        value: function renderRoutes(routes) {
            return routes.map(function (route, index) {
                return _react2.default.createElement(_Route2.default, {
                    path: route.path,
                    index: route.index,
                    component: route.component,
                    placeholder: route.placeholder,
                    initialProps: route.initialProps,
                    viewType: _constants.viewType.page,
                    exact: true,
                    key: index
                });
            });
        }

        // 渲染layout

    }, {
        key: "renderLayoutRoutes",
        value: function renderLayoutRoutes() {
            var _this2 = this;

            return _react2.default.createElement(
                _Switch2.default,
                null,
                this.state.layoutRoutes.map(function (layoutRoute, index) {
                    return layoutRoute.layout ? _react2.default.createElement(
                        _Route2.default,
                        {
                            path: layoutRoute.routes.map(function (route) {
                                return {
                                    path: route.path,
                                    index: route.index
                                };
                            }),
                            component: layoutRoute.layout, viewType: _constants.viewType.layout,
                            exact: true,
                            key: index
                        },
                        _react2.default.createElement(
                            _Switch2.default,
                            null,
                            _this2.renderRoutes(layoutRoute.routes)
                        )
                    ) : _this2.renderRoutes(layoutRoute.routes);
                })
            );
        }
    }, {
        key: "render",
        value: function render() {
            return _react2.default.createElement(
                _reactRouterRedux.ConnectedRouter,
                { history: this.props.history },
                this.renderLayoutRoutes()
            );
        }
    }]);
    return Router;
}(_react.Component);

Router.propTypes = {
    history: _propTypes2.default.object,
    routes: _propTypes2.default.array
};
Router.childContextTypes = {
    layoutStacks: _propTypes2.default.object,
    pageStacks: _propTypes2.default.object
};
exports.default = Router;
